azurelinuxagent/common/cgroupconfigurator.py

```diff
@@ -472,26 +478,96 @@ def _invoke_cgroup_operation(operation, error_message):
 
         def create_agent_cgroups(self, track_cgroups):
+            """
+            Creates and returns the cgroups needed to track the VM Agent
+            """
             def __impl():
                 cgroups = self.cgroups_api.create_agent_cgroups()
 
                 if track_cgroups:
                     # TODO: Add to tracking list
                     pass
 
+                return cgroups
+
             self._invoke_cgroup_operation(__impl, "Failed to create a cgroup for the VM Agent; resource usage for the Agent will not be tracked")
 
         def create_extension_cgroups_root(self):
+            """
+            Creates the container (directory/cgroup) that includes the cgroups for all extensions (/sys/fs/cgroup/*/walinuxagent.extensions)
+            """
             def __impl():
                 self.cgroups_api.create_extension_cgroups_root()
 
             self._invoke_cgroup_operation(__impl, "Failed to create a root cgroup for extensions; resource usage for extensions will not be tracked")
 
+        def create_extension_cgroups(self, name):
+            """
+            Creates and returns the cgroups for the given extension
+            """
+            def __impl():
+                cgroups = self.cgroups_api.create_extension_cgroups(name)
+
+                # TODO: Add to tracking list
+
+                return cgroups
+
+            self._invoke_cgroup_operation(__impl, "Failed to create a cgroup for extension '{0}'; resource usage will not be tracked".format(name))
+
+        def remove_extension_cgroups(self, name):
+            """
+            Deletes the cgroup for the given extension
+            """
+            def __impl():
+                cgroups = self.cgroups_api.remove_extension_cgroups(name)
+
+                # TODO: Add to tracking list
+
+                return cgroups
+
+            self._invoke_cgroup_operation(__impl, "Failed to delete cgroups for extension '{0}'.".format(name))
+
+        def start_extension_command(self, extension_name, command, cwd, env, stdout, stderr):
+            """
+            Starts a command (install/enable/etc) for an extension and adds the command's PID to the extension's cgroup
+            :param extension_name: The extension executing the command
+            :param command: The command to invoke
+            :param cwd: The working directory for the command
+            :param env:  The environment to pass to the command's process
+            :param stdout: File object to redirect stdout to
+            :param stderr: File pbject to redirect stderr to
+            """
+            def pre_exec_function():
+                os.setsid()
+
+                def add_to_cgroup():
+                    self.cgroups_api.add_process_to_extension_cgroup(extension_name, os.getpid())
+
+                self._invoke_cgroup_operation(add_to_cgroup, "Failed add extension '{0}' to its cgroup; resource usage will not be tracked".format(extension_name))
```

Comment:
nit: "Failed add extension" -> "Failed to add extension"

Reply:
fixed